### PR TITLE
Bugfix/mirror symlinks

### LINKED ISSUE #13722: Error when creating mirror in already existing mirror location
I ran spack twice in a row to create a mirror in the same location.  I expected it to update the mirror location with new packages.  It errored.

### Steps to reproduce the issue

```console
$ spack/bin/spack mirror create -d /usr/WS2/white238/serac/mirror --dependencies serac_devtools
$ spack/bin/spack mirror create -d /usr/WS2/white238/serac/mirror --dependencies serac_devtools
```

### Error Message

==> Adding package serac_devtools@fakeversion to mirror
==> Error: 'FileExistsError' object has no attribute 'message'

Running with -d got two stack traces in a single spack run:

Traceback (most recent call last):
  File "/usr/WS2/white238/serac/repo/uberenv_libs/spack/lib/spack/spack/mirror.py", line 506, in add_single_spec
    pkg_stage.cache_mirror(mirror_stats)
  File "/usr/WS2/white238/serac/repo/uberenv_libs/spack/lib/spack/spack/util/pattern.py", line 68, in getter
    getattr(item, self.name)(*args, **kwargs)
  File "/usr/WS2/white238/serac/repo/uberenv_libs/spack/lib/spack/spack/stage.py", line 508, in cache_mirror
    self.mirror_paths.cosmetic_path)
  File "/usr/WS2/white238/serac/repo/uberenv_libs/spack/lib/spack/spack/caches.py", line 72, in store
    os.symlink(relative_dst, cosmetic_path)
FileExistsError: [Errno 17] File exists: 'serac_devtools-fakeversion.tar.gz' -> '/usr/WS2/white238/serac/mirror/serac_devtools/serac_devtools-fakeversion.tar.gz'

==> [2019-11-13-11:04:55.099182] Successfully updated mirror in file:///usr/WS2/white238/serac/mirror
  Archive stats:
    50   already present
    0    added
    1    failed to fetch.
==> [2019-11-13-11:04:55.099393] Error: Failed downloads:
serac_devtools@fakeversion
Traceback (most recent call last):
  File "/usr/WS2/white238/serac/repo/uberenv_libs/spack/lib/spack/spack/main.py", line 715, in main
    return _invoke_command(command, parser, args, unknown)
  File "/usr/WS2/white238/serac/repo/uberenv_libs/spack/lib/spack/spack/main.py", line 457, in _invoke_command
    return_val = command(parser, args)
  File "/usr/WS2/white238/serac/repo/uberenv_libs/spack/lib/spack/spack/cmd/mirror.py", line 338, in mirror
    action[args.mirror_command](args)
  File "/usr/WS2/white238/serac/repo/uberenv_libs/spack/lib/spack/spack/cmd/mirror.py", line 324, in mirror_create
    sys.exit(1)
SystemExit: 1

### Information on your system

This includes:

 1. llnl, toss3, rzgenie
 2. Custom package.py:

from spack import *

class SeracDevtools(BundlePackage):
    """This is a set of tools necessary for the developers of Serac"""

    version('fakeversion')

    depends_on('astyle')
    depends_on('cmake')
    depends_on('cppcheck')
    depends_on('doxygen')
    depends_on('python')
    depends_on('py-sphinx')

### LINKED ISSUE #14067: `spack mirror create` creates broken symlinks
I ran `spack mirror create -d /home/razumov/eos -D -n 1 my-bundle` (where `my-bundle` is a bundle packages with a ton of dependencies). The command ran fine, but the symlinks in `/home/razumov/eos` were broken, e.g.:

```
$ pwd
/home/razumov/spack 
$ ls -l /home/razumov/eos/aida
eos/aida/aida-3.2.1.tar.gz -> ../../spack/_source-cache/archive/c5/c51da83e99c0985a7ef3e8bc5a60c3ae61f3ca603b61100c2438b4cdadd5bb2e.tar.gz
$ ls -l /home/razumov/eos/aida/../../spack/_source-cache/archive/c5/c51da83e99c0985a7ef3e8bc5a60c3ae61f3ca603b61100c2438b4cdadd5bb2e.tar.gz
ls: cannot access /home/razumov/eos/aida/../../spack/_source-cache/archive/c5/c51da83e99c0985a7ef3e8bc5a60c3ae61f3ca603b61100c2438b4cdadd5bb2e.tar.gz: No such file or directory
```

There is no `_source-cache` under `/home/razumov/spack` - but there is one under `/home/razumov/eos`:

```
$ ls /home/razumov/eos/_source-cache/archive/c5/c51da83e99c0985a7ef3e8bc5a60c3ae61f3ca603b61100c2438b4cdadd5bb2e.tar.gz
rw-r--r--. 1 razumov razumov 313562 Dec 10 14:50 /home/razumov/eos/_source-cache/archive/c5/c51da83e99c0985a7ef3e8bc5a60c3ae61f3ca603b61100c2438b4cdadd5bb2e.tar.gz
```

### Steps to reproduce the issue

```console
$ spack mirror create -d /some/path -D -n1 aida@3.2.1
$ ls /some/path/aida
```

### Error Message

None

### Information on your system

`spack` commit c627e5a6fcefb44330b4980835f0cf3a34ff2944

### PR DESCRIPTION
**Pull request by scheibelp**
(update 12/4: this PR should be rebased the commits apply fixes for thematically-related but functionally-separate problems, and I intend that the commits are properly-formatted and organized for that)

Fixes #13722 (or rather addresses the parts of it that were not addressed by #13789)
Fixes #14067
Closes #14093

* Don't try to fetch `BundlePackage`s
* Don't fetch patches if they were already added to the mirror
* #13789 started calculating a relative path, relative to another relative path. This can fail based on the working directory, so the intended reference point for the relative path is created explicitly as the absolute path
* Allow repeated invocations of `spack mirror create` on the same mirror directory (in develop this currently reports errors)

### CODE REVIEW

**Comment by scheibelp** [comment]
@tgamblin this particular PR should be rebased: I have formatted all the commit messages.

**Comment by alalazo** [comment]
> this particular PR should be rebased (vs. squashed): I have formatted all the commit messages.

Question related to commit messages: can you add a  handle to this PR in each commit if this is not squashed? Asking because it's quite useful when you read code to annotate it and eventually read the PRs associated with the lines of interest.

**Comment by sknigh** [comment]
@scheibelp I don't know if the other PRs fix the long timeout issue, but I tweaked this branch to get a better behavior. I added `--connect-timeout` to the curl arguments. It only exits when an initial connection cannot be established within a 15 second limit.

``` patch
diff --git a/lib/spack/spack/fetch_strategy.py b/lib/spack/spack/fetch_strategy.py
index 409cc334e..a6956b7d4 100644
--- a/lib/spack/spack/fetch_strategy.py
+++ b/lib/spack/spack/fetch_strategy.py
@@ -324,6 +324,7 @@ def fetch(self):
             '-D',
             '-',  # print out HTML headers
             '-L',  # resolve 3xx redirects
+            '--connect-timeout', '15',
             self.url,
         ]
 
diff --git a/lib/spack/spack/mirror.py b/lib/spack/spack/mirror.py
index ceb53801c..3a863188a 100644
--- a/lib/spack/spack/mirror.py
+++ b/lib/spack/spack/mirror.py
@@ -499,7 +499,7 @@ def add_single_spec(spec, mirror_root, mirror_stats):
     tty.msg("Adding package {pkg} to mirror".format(
         pkg=spec.format("{name}{@version}")
     ))
-    num_retries = 3
+    num_retries = 1
     while num_retries > 0:
         try:
             with spec.package.stage as pkg_stage:
```

**Comment by alalazo** [comment]
@sknigh See #13881 for a PR that adds the same flag.

File: lib/spack/spack/stage.py [hunk 1]
```diff
@@ -494,6 +494,16 @@ def cache_local(self):
...
     def cache_mirror(self, stats):
         """Perform a fetch if the resource is not already cached"""
+        if isinstance(self.default_fetcher, fs.BundleFetchStrategy):
+            # BundleFetchStrategy has no source to fetch. The associated
+            # fetcher does nothing but the associated stage may still exist.
+            # There is currently no method available on the fetcher to
+            # distinguish this ('cachable' refers to whether the fetcher
+            # refers to a resource with a fixed ID, which is not the same
+            # concept as whether there is anything to fetch at all) so we
+            # must examine the type of the fetcher.
```
Comment (alalazo) [inline lib/spack/spack/stage.py]: :100: for the clear comment

File: lib/spack/spack/caches.py [hunk 2]
```diff
@@ -67,10 +67,11 @@ def symlink(self, mirror_ref):
         storage location."""
 
         cosmetic_path = os.path.join(self.root, mirror_ref.cosmetic_path)
+        storage_path = os.path.join(self.root, mirror_ref.storage_path)
         relative_dst = os.path.relpath(
-            mirror_ref.storage_path,
+            storage_path,
             start=os.path.dirname(cosmetic_path))
-        if not os.path.exists(cosmetic_path):
+        if not os.path.lexists(cosmetic_path):
```
[the remark this answers is not in the document]
Comment (scheibelp) [reply lib/spack/spack/caches.py]: Sure and in fact we might as well just remove it if it exists and overwrite it (which is to say, I don't think it's required to add logic to make sure the link points where we want it to, we can just rewrite it all the time).

**Comment by scheibelp** [comment]
This is failing because as part of addressing https://github.com/spack/spack/pull/13908#discussion_r356466494 I invoke `os.remove` on the symlink which actually deletes the file that it points to. Oddly `os.unlink` appears to fail in this way also, so I'm not sure if there is a way to remove the symlink properly.

I can either:

* hack the ability to delete symlinks for existing files into Spack (e.g. by temporarily moving the symlink target to a temporary directory)
* examine the link but never delete it (and inform the user of the issue)

The latter does not seem very reasonable to me so I'll be thinking about how to do the first one.

**Comment by alalazo** [comment]
@scheibelp I think `os.unlink` does what you want on linux (remove the link without deleting what it points to) but:

1. The [documentation](https://docs.python.org/3/library/os.html#os.unlink) contradicts the current behavior (it says it has the same semantic a `os.remove` which is not always true)
2. The behavior might be OS/platform specific

Another alternative might be to return early if the link `os.exists`:
```python
if os.path.exists(cosmetic_path):
    return
```
which means it's not broken (so basically rule out every exception to the happy code path at the beginning of the method).

**Comment by scheibelp** [comment]
Thanks for the suggestion!

> I think os.unlink does what you want on linux

I encountered this problem on Mac OS so that might be why it comes up for me

> Another alternative might be to return early if the link os.exists:

I agree that checking both `os.path.lexists` and `os.path.exists` should resolve the issue for all platforms. I'll be able to fix that this evening.

**Comment by scheibelp** [comment]
https://github.com/spack/spack/pull/13908#issuecomment-564879464 worked for me but to clarify:

I asserted in https://github.com/spack/spack/pull/13908#issuecomment-564840630 that `os.unlink` was deleting the target of a symlink. That's not true: it turns out some tests were creating a file vs. a symlink at `cosmetic_path`, and `os.unlink` will delete the provided path if it refers to a file (vs. a symlink), which was causing the error.

So at this point all requests are addressed.

**Pull request merged**